src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Managed.cs

```diff
@@ -90,26 +94,49 @@ public void Abort()
                 AddWebSocketHeaders(request, secKeyAndSecWebSocketAccept.Key, options);
 
                 // Create the handler for this request and populate it with all of the options.
-                handler = new SocketsHttpHandler();
-                handler.Credentials = options.Credentials;
-                handler.Proxy = options.Proxy;
-                handler.CookieContainer = options.Cookies;
-                handler.SslOptions.RemoteCertificateValidationCallback = options.RemoteCertificateValidationCallback;
-                if (options._clientCertificates?.Count > 0) // use field to avoid lazily initializing the collection
+                // Try to use a shared handler rather than creating a new one just for this request, if
+                // the options are compatible.
+                if (options.Credentials == null &&
+                    options.Proxy == null &&
+                    options.Cookies == null &&
+                    options.RemoteCertificateValidationCallback == null &&
+                    options._clientCertificates?.Count == 0)
+                {
```

Comment:
Q.: should we also check for impersonation? I ask because the docs for From [WinHttpOpen](https://msdn.microsoft.com/en-us/library/windows/desktop/aa384098(v=vs.85).aspx): "an application that makes requests on behalf of multiple users should create a separate session for each user, so as not to share user-specific cookies and authentication state."

Reply:
That's probably true for WinHTTP based code.  But websockets aren't using WinHTTP anymore. So, whatever state might be shared within SocketsHttpHandler needs to be considered.

Reply:
The call to `WinHttpOpen` on Windows still can happen from `SocketsHttpHandler` when `UseProxy` is true (the default) and `Proxy` is null, see https://github.com/dotnet/corefx/blob/8ce711e455f33f7a4bcb9cdb0c5374984e11035e/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs#L71

Reply:
>or do you mean that for the proxy usage it doesn't matter?

I think this needs more thought.  

It seems that `ClientWebSocketOptions` only has a `Proxy` property and not any boolean UseProxy property.
https://github.com/dotnet/corefx/blob/master/src/System.Net.WebSockets.Client/ref/System.Net.WebSockets.Client.cs#L27

So, what is the behavior if ClientWebSocketOptions.Proxy is null?  Does it mean never use a proxy?  Or does it mean always use the system default proxy?  The system default proxy settings are queried at the platform level (registry settings-Windows, environment variable-Linux). After doing so, there still might not be a proxy for a particular request URI.

Reply:
> So, what is the behavior if ClientWebSocketOptions.Proxy is null? Does it mean never use a proxy? Or does it mean always use the system default proxy? 

This was going to be my question. We seem to already differ here from netfx.

Reply:
I don't remember the behavior on .NET Framework regarding this. Might need to be tested.

But if the behavior is such that .Proxy==null means never use a proxy, then the use of SocketsHttpHandler needs to be changed to set the .UseProxy property to false if .Proxy==null.

Reply:
Just from code inspection, it looks like options.Proxy defaults to non-null, set to a default proxy instance, and then HWR.Proxy is just see to whatever options.Proxy is at the time.

Reply:
>it looks like options.Proxy defaults to non-nul

Ok. In that case, it is following the pattern of the original HttpWebRequest API which didn't have a separate UseProxy Boolean.  The default for HttpWebRequest.Proxy is non-null and points to the "system default proxy" object.  Setting HttpWebRequest.Proxy to null means never use a Proxy.

HttpClient API changed that pattern and added a boolean 'UseProxy'.

So, based on this, we should make sure that the implementation of ClientWebSocketOptions.Proxy is such that having a null value means to set the 'UseProxy' of SocketsHttpHandler to false.

Reply:
Ok. I'll add a commit for that.

Reply:
That sounds much better!

I guess that this will make ClientWebSocket for the default case much better, if I have a chance I will measure it before and after the change on the WCF stress.

Reply:
> I guess that this will make ClientWebSocket for the default case much better

Unfortunately, no.  netfx defaults to a non-null proxy that represents the system defaults, so I've made the netcoreapp one do so as well.  But you can at least explicitly set Proxy to null now, and in doing so we'll set UseProxy on the handler to false, so there's now at least a way to avoid using the default system proxy and thus avoid doing anything with WinHttpOpen.

Reply:
I see. 

Looking at [WCF usage](https://github.com/dotnet/wcf/blob/a2c95ad9e5cb2fff465655ae234454cf6f1c0f83/src/System.Private.ServiceModel/src/System/ServiceModel/Channels/CoreClrClientWebSocketFactory.cs#L20) they will not be hitting the path of the shared one, but, they will be using the DefaultWebProxy.Instance so no more WinHttpOpen in their case.

---
In reply to: [174168555](https://github.com/dotnet/corefx/pull/27994#discussion_r174168555) [](ancestors = 174168555)

Reply:
> but, they will be using the DefaultWebProxy.Instance so no more WinHttpOpen in their case.

I would still expect WinHttpOpen there.  If the proxy is DefaultWebProxy.Instance, then SocketsHttpHandler.UseProxy will be true and Proxy will be null, which means SocketsHttpHandler on Windows will end up looking up a default proxy using WinHttpOpen.

Reply:
(If they could set Proxy to null or to some other IWebProxy instance, then WinHttpOpen would be skipped.)

Reply:
@mconnew @KKhurin FYI: the information above, regarding setting the proxy, should be considered when moving to `SocketsHttpHandler`.

Reply:
There was some reason why we didn't enable proxy usage with WebSockets initially, but if that's fully working, we absolutely will always be setting the Proxy. We either set it to an instance of WebProxy which represents the system proxy, or an instance of WebProxy for a specific proxy, or the user can specify null. But I also intend to switch to using SocketsHttpHandler directly to do the WebSockets upgrade handshake and then create a WebSocket instance from the stream. This is because the WebSockets api doesn't expose everything from the HTTP handshake and WCF needs this. For example, we actually use the response headers but that's not available with ClientWebSocket.